### ISSUE TITLE
Make logging more modular

### ISSUE THREAD

**Issue by Th0rstenf** [Th0rstenf/Rocksniffer-to-Streamer.bot-extension#60]
The debug should only output the general configuration and errors that come up during the plugin execution.

Error messages that just evaluate all program flow status and steps should not be active outside of development

**Comment by kozaka-tv** [comment]
Idk...is there no such thing like Log Level in c++?

But actually I think, we could introduce a global variable like logLevel. Default INFO level. Almost nothing is written out on this level. On DEBUG level, we should see the changes, like current scene, Sniffer status, calculated values, etc.
Maybe this way we could just change the log level also in SB, what in some cases would be nice to have.

**Comment by Th0rstenf** [comment]
I tend to disagree. If we use the SB levels, we will just create huge log files as many people have SB configured as debug without even knowing.
 Logging the configuration and occuring errors should be totally sufficient, as any error that is not shown should be a wrong configuration.

Though we could also do that, use CPH.LogInfo for the configuration and LogDebug for everything else and add a remark to the readme that the debug configuration will flood the log excessively.
I still don't see the point though, as all errors i've seen in the wild were due to misconfiguration, which is already logged.
In development I really don't look at those messages at all, as the mock object already logs CPH interface calls.

**Comment by kozaka-tv** [comment]
No, I did not know that SB has leveles. Just I saw that my SB loggs out debug stuffs...this is what you mean under "SB configured as debug without even knowing" I assume :) But lot of people not even know that there is a log! :)

Errors should be logged out always, no question. What I expect from a log, that I wanna see, that it is working/running (small, rare info logs in every minutes or so...idk...to see that it is running). Or nothing. Also good...If it works, then works.

And in case I enable debug level, debug logs with lot of Information should apear. It could be, that later we ask users to enable debug, or attach the log because of a bug we want to analyze. If there is no option to have detailed information about what the app does, it could be hard to fix a bug.

Hm... where is set in SB the log levele?

In our code, I think logDebug variable is nice. Just we should have this as default false in the releases.

**Comment by Th0rstenf** [comment]
Agreed, but we can't influence that here. As you recognized, debug is default. 
Usually this is probably no issue, most actions in streamer.bot are not happening every second. For us it's pretty harsh.

CPH offers those methods directly:

void LogInfo(string logLine);  --> Maybe this for Konfiguration 
void LogWarn(string logLine);  --> Not sure we need this?
void LogDebug(string logLine); --> Obvious errors, caught exceptions, etc
void LogVerbose(string logLine); --> Complete information

This way the default debug level would already give us the most relevant information, and in rare edge cases we could ask people to switch to verbose?

**Comment by kozaka-tv** [comment]
Yeah, good idea! As the debug is default, we should use as it were info level. And use verbose as debug. If SB will change the default level to info, then we could change this if needed.

Now, if you check, SB loggs out every second some lines because of the timed action. So, this is acceptable for them.

I can do this. And I will then read the log level from a config or something for the mock. Or we could just have it hardcoded also.

**Comment by Th0rstenf** [comment]
As there is no error evel, AND it is default, weet should treat debug as
error

Kozaka ***@***.***> schrieb am Mi., 22. März 2023, 12:14:

**Comment by Th0rstenf** [comment]
> Yeah...or we just throw an exception and that's it. I will create a new
suggestion for LogError, because I think this is missing.

No, we will treat LogDebug as an error output. All the information you are
logging is really verbose, this is much beyond the scope of debug and not
acceptable for users that will have it active per default anyway.

Am Mi., 22. März 2023 um 13:21 Uhr schrieb Kozaka ***@***.***

**Comment by kozaka-tv** [comment]
I will do regadring logging:
1) Introduce LogLevel in our app to get rid of SB-s stuff. This is as default Info.
2) CPH.Log calls will be called according to the current internal Log level (not SB log level, as right now we can not get the current SB log level).
3) proper CPH.Log calls will be introduced. Considered default SB log level is Debug (means almost all of the LogDebug calls will go to LogVerbose, but some...just a couple, goes to Info, Warn or Error.)

In case, SB sets default log level to Info + LogLevel in SB could be get, this "override" could be removed, and we could just keep the CPH.Log calls as they are.

Use cases:
1)
SB Log level = Debug
APP Log level = Info
LogVerbose -> Not
LogDebug --> Not
LogInfo --> Logged
LogWarn --> Logged
LogError --> Logged

2)
SB Log level = Verbose
APP Log level = Debug
LogVerbose -> Not
LogDebug --> Logged
LogInfo --> Logged
LogWarn --> Logged
LogError --> Logged

3)
SB Log level = Verbose
APP Log level = Verbose
LogVerbose -> Logged
LogDebug --> Logged
LogInfo --> Logged
LogWarn --> Logged
LogError --> Logged

4) In case, APP will Verbose, but SB "blocks"
SB Log level = Debug
APP Log level = Verbose
LogVerbose -> Not
LogDebug --> Logged
LogInfo --> Logged
LogWarn --> Logged
LogError --> Logged

I think this could work! And then as in SB Debug is default and in the App Info, there will be just some important logs, warnings, errors.

**Comment by kozaka-tv** [comment]
Long story In short: I will not do this config stuff in Programm.cs, just shift down some LogDebug to LogVerbose.
---------------

I spoke a bit with SB guys, and there will be probably no option to use LogError, and somehow this log config stuff just does not worked (had problem with static calls from non static what I could not solve).

But the whole stuff is not so important to have a very sophisticated logging I think. We have just to have like this.

In 0.1.19. Default log level is Debug so I think we have just have to shift a bit deeper some logging like:

- LogVerbose = Normal Debug level: states, variable values, smaller details regarding processes, not for normal user stuffs. Ment to dev, for us.
- LogDebug = Normal Info level: Just some common, major informations like config values, scene changes
- LogInfo = Should be not really used, as Debug level is info

And about warnings and exceptions. I think We could use warning in case like some configurations are not ok, or like Broadcasting software is not configured. In case exception (can not connect to Sniffer, parsing errors, etc), we should write out a Warning (as LogError will be most likely not enabled for us what I think is bad), and throw a new, or rethrow the old exception, to have in log more visible error messages. What I do not like in this situation, is that, because LogError does not exists, and Debug is default, there is a lot of log entry, but the real problems, the really important log rows are almost hidden. Here is an example for a Warning:

![image](https://user-images.githubusercontent.com/4964462/228796882-8ebfac75-453f-44cb-a3aa-2422f59be6fd.png)

I think an Exception would be better in this case for the user and for us:

![image](https://user-images.githubusercontent.com/4964462/228797179-4df9b47b-9579-4d8b-9d20-fad8bb8d42ab.png)

Weird situation :)

---

In case, Info log level will be default in SB, we could shift back the logs to they normal state:
- LogVerbose = variable values, states
- LogDebug = logs about processes like, where am I, what I do
- LogInfo = Config, scene changes, major stuffs
- LogWarning = Warnings, bigger problems what user should not avoid.

https://ideas.streamer.bot/posts/708/set-default-log-level-to-information

**Comment by kozaka-tv** [comment]
And, in case we got some problem (like could not connect to sniffer), we can not do anything. This SB solution differs from the original Wraths Switcher solution in that way, that this program runs only once, but will be restarted in each seconds. So, if there is an error, the actual process should be just stopped with an error message. Then user checks log, and see, that Sniffer connection is not possible, and do fix. Or turn off the switcher.

For us, debugging should be possible with IDE or Verbose log in SB.

**Comment by kozaka-tv** [comment]
**Or, what if we do logging as normal, use Info, debug, verbose** and if default log level stays DEBUG in SB, we just ask the users in readme to use set Log level to INFO if the longer log files are disturbing? Then we do not need to bother with the topic too much!
